File: engine/src/main/java/de/hanno/hpengine/engine/model/AbstractModel.java
package de.hanno.hpengine.engine.model;

import com.carrotsearch.hppc.FloatArrayList;
import com.carrotsearch.hppc.IntArrayList;
import de.hanno.hpengine.engine.model.material.MaterialFactory;
import de.hanno.hpengine.engine.transform.AABB;
import de.hanno.hpengine.engine.transform.SimpleSpatial;
import de.hanno.hpengine.engine.transform.Transform;
import de.hanno.hpengine.engine.graphics.buffer.Bufferable;
import de.hanno.hpengine.engine.model.material.Material;
import de.hanno.hpengine.engine.transform.Spatial;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class AbstractModel<T extends Bufferable> extends SimpleSpatial implements Model<T>, Spatial {
    protected final List<Mesh<T>> meshes;
    protected int triangleCount;
    protected IntArrayList[] meshIndices;

    public AbstractModel(List<Mesh<T>> meshes) {
        this.meshes = meshes;
    }

    protected void init(MaterialFactory materialFactory) {
        for (int i = 0; i < meshes.size(); i++) {
            Mesh mesh = meshes.get(i);
            AABB meshMinMax = mesh.getMinMax();
//            StaticMesh.calculateMinMax(getMinMax().getMin(), getMinMax().getMax(), meshMinMax);
            StaticMesh.calculateMinMax(getMinMaxProperty().getMin(), getMinMaxProperty().getMax(), meshMinMax);
        }
    }

    public void setMaterial(Material material) {
        for (Mesh mesh : meshes) {
            mesh.setMaterial(material);
        }
    }

    public int getTriangleCount() {
        return triangleCount;
    }

    public float[] getVertexBufferValuesArray() {
        FloatArrayList floatList = new FloatArrayList();
        for(Mesh mesh : meshes) {
            floatList.add(mesh.getVertexBufferValuesArray());
        }
        return floatList.toArray();
    }

    public List<T> getCompiledVertices() {
        List<T> vertexList = new ArrayList<>();
        for(Mesh mesh : meshes) {
            vertexList.addAll((Collection<? extends T>) mesh.getCompiledVertices());
        }
        return vertexList;
    }

    public int[] getIndices() {
        IntArrayList intList = new IntArrayList();
        for(Mesh mesh : meshes) {
            intList.add(mesh.getIndexBufferValuesArray());
        }
        return intList.toArray();
    }

    @Override
    public AABB getMinMax(Transform transform) {
        return super.getMinMaxWorld(transform);
    }

    public List<Mesh<T>> getMeshes() {
        return meshes;
    }

    public IntArrayList[] getMeshIndices() {
        return meshIndices;
    }

}

File: engine/src/main/java/de/hanno/hpengine/engine/model/AbstractModel.kt
package de.hanno.hpengine.engine.model

import de.hanno.hpengine.engine.graphics.renderer.pipelines.IntStruct
import de.hanno.hpengine.engine.model.material.Material
import de.hanno.hpengine.engine.transform.AABB
import de.hanno.hpengine.engine.transform.SimpleSpatial
import de.hanno.hpengine.engine.transform.Spatial
import de.hanno.struct.Struct
import de.hanno.struct.copyTo
import org.joml.Matrix4f


@JvmOverloads
fun <T: Struct> de.hanno.struct.StructArray<T>.copyTo(target: de.hanno.struct.StructArray<T>, offset: Int, rewindBuffers: Boolean = true) {
    buffer.copyTo(target.buffer, rewindBuffers, slidingWindow.sizeInBytes * offset)
}

abstract class AbstractModel<T>(final override val meshes: List<Mesh<T>>,
                                material: Material) : SimpleSpatial(), Model<T>, Spatial {

    final override val meshIndexCounts = meshes.map { it.indexBufferValues.size }
    val meshIndexSum = meshIndexCounts.sum()

    override var indices = de.hanno.struct.StructArray(meshIndexSum) { IntStruct() }.apply {
        var offsetPerMesh = 0
        meshes.forEach { mesh ->
            mesh.indexBufferValues.copyTo(this, offset = offsetPerMesh)
            offsetPerMesh += mesh.indexBufferValues.size
        }
    }
    override var triangleCount: Int = meshes.sumBy { it.triangleCount }
    override val uniqueVertices: List<T> = meshes.flatMap { it.vertices }

    override var material: Material = material
        set(value) {
            meshes.forEach { it.material = value }
            field = value
        }

    override fun getBoundingVolume(transform: Matrix4f): AABB = super<SimpleSpatial>.getBoundingVolume(transform)
    abstract fun calculateBoundingVolume(): AABB
}
